opm/autodiff/AquiferCarterTracy.hpp

```diff
@@ -0,0 +1,418 @@
+#include <opm/parser/eclipse/EclipseState/Aquancon.hpp>
+#include <opm/autodiff/BlackoilAquiferModel.hpp>
+#include <opm/common/OpmLog/OpmLog.hpp>
+#include <opm/common/utility/numeric/linearInterpolation.hpp>
+#include <opm/core/props/BlackoilPhases.hpp>
+
+
+#include <opm/material/densead/Math.hpp>
+#include <opm/material/densead/Evaluation.hpp>
+#include <opm/material/fluidstates/BlackOilFluidState.hpp>
+
+#include <string>
+#include <memory>
+#include <vector>
+#include <algorithm>
+#include <map>
+#include <cassert>
+
+namespace Opm
+{
+
+    template<typename TypeTag>
+    class AquiferCarterTracy
+    {
+
+        public:
+            typedef BlackoilModelParameters ModelParameters;
+
+            typedef typename GET_PROP_TYPE(TypeTag, Grid) Grid;
+            typedef typename GET_PROP_TYPE(TypeTag, GridView) GridView;
+            typedef typename GET_PROP_TYPE(TypeTag, Simulator) Simulator;
+            typedef typename GET_PROP_TYPE(TypeTag, FluidSystem) FluidSystem;
+            typedef typename GET_PROP_TYPE(TypeTag, Indices) BlackoilIndices;
+            typedef typename GET_PROP_TYPE(TypeTag, IntensiveQuantities) IntensiveQuantities;
+            typedef typename GET_PROP_TYPE(TypeTag, MaterialLaw) MaterialLaw;
+            typedef typename GridView::template Codim<0>::Entity Element;
+            typedef typename GET_PROP_TYPE(TypeTag, ElementContext) ElementContext;
+
+            static const int numEq = BlackoilIndices::numEq;
+            typedef double Scalar;
+
+            typedef DenseAd::Evaluation<double, /*size=*/numEq> Eval;
+            typedef Opm::BlackOilFluidState<Eval, FluidSystem> FluidState;
+
+            static const auto waterCompIdx = FluidSystem::waterCompIdx;
+            static const auto waterPhaseIdx = FluidSystem::waterPhaseIdx;
+
+
+
+            explicit AquiferCarterTracy( const AquiferCT::AQUCT_data& params, const Aquancon::AquanconOutput& connection,
+                                         const Scalar gravity, const Simulator& ebosSimulator                               )
+            :  ebos_simulator_ (ebosSimulator),
+              aquiferID_ (params.aquiferID),
+              inftableID_ (params.inftableID),
+              pvttableID_ (params.pvttableID),
+              phi_aq_ (params.phi_aq), //
+              d0_ (params.d0),
+              C_t_ (params.C_t), //
+              r_o_ (params.r_o), //
+              k_a_ (params.k_a), //
+              c1_ (params.c1),
+              h_ (params.h), //
+              theta_ (params.theta), //
+              c2_ (params.c2), //
+              aqutab_td_ (params.td),
+              aqutab_pi_ (params.pi),
+              pa0_ (params.p0),
+              gravity_ (gravity),
+              p0_defaulted_ (params.p0_defaulted)
+            {
+                init_quantities(connection);
+            }
+
+            inline void assembleAquiferEq(Simulator& ebosSimulator, const SimulatorTimerInterface& timer)
```

Comment:
Since you have `ebos_simulator_` as a member variable, `ebosSimulator` is probably not needed here.